### PR TITLE
HQ support for the Reminders App

### PR DESCRIPTION
**Pull request by kaapstorm**
## Summary

Jira: [SC-953](https://dimagi-dev.atlassian.net/browse/SC-953)

This PR implements the HQ side of the work to support the Reminders App.

Case List and Form settings include a text field under "Settings" > "Advanced" labeled "Session Endpoint ID". The value is used to generate an `<endpoint>` node in the app's **suite.xml** file.

This PR includes a couple of commented-out "**TO DO**"s to hide/disable the new functionality if the app is using a version of CommCare that does not support session endpoints. I am keen to hear feedback on whether people think this is useful to uncomment once the upcoming version of CommCare is released, or whether I should drop the commented-out code.

:fish: :blowfish: :tropical_fish: :octopus: Easier to review by commit.

fyi @shubham1g5 

## Feature Flag

'Enable session endpoints'

## Product Description

HQ support for the Reminders App. 

## Safety Assurance

- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I am certain that this PR will not introduce a regression for the reasons below

### Automated test coverage

PR includes automated test coverage. If you feel coverage is insufficient, please let me know.

### QA Plan

This branch is currently deployed to staging for review.

### Safety story

* Unit test coverage
* Tested locally
* Tested on Staging

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### CODE REVIEW

File: corehq/toggles.py
Comment (proteusvacuum) [inline corehq/toggles.py:1550]: nit: extra newline here

File: corehq/apps/app_manager/suite_xml/sections/endpoints.py
Comment (proteusvacuum) [inline corehq/apps/app_manager/suite_xml/sections/endpoints.py:69]: I'm maybe not completely understanding how this works, but how is this session variable populated in the reminders app workflow? Doesn't that mean that a user will have had to already navigate to the case they want in CommCare before navigating the reminders app? Or is there some other way this is populated? 

The spec seems to suggest using an `<argument>` whose name is the value here, which I assume is something that the reminders app would pass in when this callout is triggered. 
```
<argument id=”person_case_id”/>
...
<datum id=”case_id” value=”$person_case_id”>
````

Comment (esoergel) [reply corehq/apps/app_manager/suite_xml/sections/endpoints.py:69]: that sounds in line with Shubham's comment [here](https://dimagi-dev.atlassian.net/browse/SC-953?focusedCommentId=115731)

File: corehq/toggles.py
Comment (esoergel) [inline corehq/toggles.py:1570]: Know if this should be `TAG_SOLUTIONS_LIMITED`, `OPEN`, or `CONDITIONAL`?

File: corehq/apps/app_manager/templates/app_manager/partials/session_endpoints.html
Comment (esoergel) [inline corehq/apps/app_manager/templates/app_manager/partials/session_endpoints.html:20]: Know if this needs to be a valid slug or anything?  Should we slap some html5 validation on it?

File: corehq/apps/app_manager/tests/test_session_endpoints.py
Comment (orangejenny) [inline corehq/apps/app_manager/tests/test_session_endpoints.py:1]: Minor naming convention: I'd call this file `test_suite_session_endpoints`.

File: corehq/apps/app_manager/feature_support.py
Comment (orangejenny) [reply corehq/apps/app_manager/feature_support.py:185]: Also if 2.51 isn't created, you can create it and mark it superuser-only.

File: corehq/apps/app_manager/suite_xml/generator.py
Comment (esoergel) [reply corehq/apps/app_manager/suite_xml/generator.py:89]: @orangejenny do you have strong opinions about this?

Comment (orangejenny) [reply corehq/apps/app_manager/suite_xml/generator.py:89]: I'd favor consistency and keep the code as it is.

File: corehq/apps/app_manager/suite_xml/sections/endpoints.py
Comment (kaapstorm) [reply corehq/apps/app_manager/suite_xml/sections/endpoints.py:69]: Yeah. This code is wrong. Shubham's comment is spot on, and clarifies the problem, and how to fix it. We need to have an `<argument>` for each case the form needs.

File: corehq/apps/app_manager/templates/app_manager/partials/session_endpoints.html
Comment (kaapstorm) [reply corehq/apps/app_manager/templates/app_manager/partials/session_endpoints.html:23]: This might be worth confirming with @shubham1g5 My understanding is that we want (wanted?) to be able to offer more than one in the future. (?)

File: corehq/toggles.py
Comment (kaapstorm) [reply corehq/toggles.py:1570]: I'm not sure. But I bet @karmuno knows.

File: corehq/apps/app_manager/templates/app_manager/partials/session_endpoints.html
Comment (kaapstorm) [reply corehq/apps/app_manager/templates/app_manager/partials/session_endpoints.html:20]: A slug sounds right.

File: corehq/apps/app_manager/suite_xml/generator.py
Comment (esoergel) [reply corehq/apps/app_manager/suite_xml/generator.py:89]: Fair enough.

File: corehq/apps/app_manager/feature_support.py
Comment (esoergel) [reply corehq/apps/app_manager/feature_support.py]: 2.51 is out on staging and prod now, so I uncommented this and a similar one in the form template

File: corehq/toggles.py
Comment (esoergel) [reply corehq/toggles.py]: Tagged it "limited use" for now - it's an easy thing to change later as desired.

File: corehq/apps/app_manager/templates/app_manager/partials/session_endpoints.html
Comment (shubham1g5) [reply corehq/apps/app_manager/templates/app_manager/partials/session_endpoints.html]: We will only have one session endpoint per form/module. Though we definitely want to have multiple session endpoints overall for different screens of an app

Comment (esoergel) [reply corehq/apps/app_manager/templates/app_manager/partials/session_endpoints.html]: Thanks - I modified the code to reflect this.

File: corehq/apps/app_manager/tests/test_suite_session_endpoints.py
Comment (esoergel) [inline corehq/apps/app_manager/tests/test_suite_session_endpoints.py:17]: @shubham1g5 this test suite contains a few examples of generated suite files - a registration form, followup form, and a case list.  Do you have any more complex examples that would be useful to add?  Perhaps weirder scenarios like child case lists, for example.  I'm not confident all edge-cases are handled appropriately.

File: corehq/apps/app_manager/suite_xml/sections/endpoints.py
Comment (kaapstorm) [inline corehq/apps/app_manager/suite_xml/sections/endpoints.py:52]: <3

**Review by kaapstorm — COMMENTED** [review]
I can't approve this PR because I opened it. But please consider it approved. This is looking great.

File: corehq/apps/app_manager/tests/test_suite_session_endpoints.py
Comment (shubham1g5) [reply corehq/apps/app_manager/tests/test_suite_session_endpoints.py:17]: This looks great, I agree that it would be nice to add an example for parent child workflow. [This form](https://staging.commcarehq.org/a/another-upstream/apps/view/b1e763d627164bccb1359585e0a5b7f7/form/05e6f3afe12c4048a7b26bd46963fd66/source/#form/the_show_is_in_case_name) in particular looks like a good example to cover. The suite file for it defines the form entry with the datums - 

````
 <session>
      <datum id="parent_id" nodeset="instance('casedb')/casedb/case[@case_type='song'][@status='open']" value="./@case_id" detail-select="m2_case_short"/>
      <datum id="case_id" nodeset="instance('casedb')/casedb/case[@case_type='show'][@status='open'][index/parent=instance('commcaresession')/session/data/parent_id]" value="./@case_id" detail-select="m3_case_short" detail-confirm="m3_case_long"/>
    </session>
````

so the session endpoint should define 2 arguments for this aka `$parent_id` and `$case_id`

Comment (esoergel) [reply corehq/apps/app_manager/tests/test_suite_session_endpoints.py:17]: Great, just added a couple of examples for that.   Let me know if anything else comes up that should be tested explicitly.

**Review by esoergel — APPROVED** [review]
AFAIK this PR is now ready to go.  I'm gonna hold off on merge to give @shubham1g5 time to try it out on staging, but any reviewers waiting for stability can now take another pass.

File: corehq/apps/app_manager/suite_xml/sections/endpoints.py
Comment (proteusvacuum) [inline corehq/apps/app_manager/suite_xml/sections/endpoints.py:62]: Is it always obvious to the user what this id will be? I'm assuming for advanced modules that this will be something like `case_id_new_{case type}_{action index}`. Is that correct, or have I misunderstood what this is actually doing?

Comment (esoergel) [reply corehq/apps/app_manager/suite_xml/sections/endpoints.py:62]: I'm actually not sure what happens on the other side of this equation.  The reminders app (or other future usage of this feature) _will_ need to know how to fill out this information, but I'm not certain how they do that.  It should be the same as how these data are made available in the session.  Do you have any thoughts around how to make it clearer?

Comment (proteusvacuum) [reply corehq/apps/app_manager/suite_xml/sections/endpoints.py:62]: Got it. I guess if you are using advanced modules you are on your own anyway, and the format for the case ids are pretty well documented. I had been under the impression that the argument ID was something that the user would configure (so they would remain static if anything else about the app changes), but sounds like this happens automatically. I don't understand this feature well enough to know if that is an issue or not, so can leave it up to you.

Comment (esoergel) [reply corehq/apps/app_manager/suite_xml/sections/endpoints.py:62]: Ah, gotcha - I think you're thinking of the endpoint ID.  That's an attribute of the `<endpoint />` node wrapping this.  In an earlier version of this PR, that was also used as the argument ID, but Shubham pointed out the argument ID should match the datum ID.  Honestly I don't have my head wrapped around exactly how that's used, but my understanding is the argument ID is to be used by the other app to figure out how to fill out the stack.

**Pull request merged**
